### ISSUE TITLE
Write correct spatial metadata to Storage Units

### ISSUE THREAD

**Issue by omad** [opendatacube/datacube-core#21]
As reported by Josh @sixy6e :

The spatial metadata in the sample files is a little off. Here's an ipython notebook documenting the problem.

https://github.com/sixy6e/my_code/blob/master/Python/notebooks/nc_metadata_tests.ipynb

The current spatial extents exclude the east-most and south-most lines of pixels, instead of including them. Correcting this should fix both the extents and the pixel sizes.

**Comment by omad** [comment]
It appears that at least part of the problem is the small dtype size **float32** that latitudes and longitudes are stored in which leads to rounding errors. I'll change them to **float64** and see how that helps.

``` python
>>> a = (numpy.arange(4000)*0.00025 + 150).astype(numpy.float32)
>>> a[400:410]-a[399:409]
array([ 0.0002594 ,  0.00024414,  0.00024414,  0.0002594 ,  0.00024414,
    0.0002594 ,  0.00024414,  0.00024414,  0.0002594 ,  0.00024414], dtype=float32)

>>> a = (numpy.arange(4000)*0.00025 + 150).astype(numpy.float64)
>>> a[400:410]-a[399:409]
array([ 0.00025,  0.00025,  0.00025,  0.00025,  0.00025,  0.00025,
        0.00025,  0.00025,  0.00025,  0.00025])

```

**Comment by sixy6e** [comment]
Good pickup, the lon/lat units will need the extra precision more than **most** meters based projections, but probably ideal to store all co-ordinate based info as **float64**.

**Comment by omad** [comment]
Hiya Josh,
I think that this should be fixed now. Could you please have a look at the files in `/short/v10/dra547/tmp/7/` and if they look alright to you, close this issue.

They're only a single timeslice, but that makes no difference to the spatial extents.

**Comment by sixy6e** [comment]
``` python
>>> import h5py
>>> from pprint import pprint
>>> fname = 'LANDSAT_5_TM_134.0_-31.0_2015-09-29T23-09-40.000000.nc'
>>> fid = h5py.File(fname, 'r')
>>> pprint(fid.attrs.items())
[(u'geospatial_lon_min', array([ 134.])),
 (u'geospatial_lat_units', 'degrees_north'),
 (u'spatial_coverage', '1.000000 degrees grid'),
 (u'geospatial_lon_units', 'degrees_east'),
 (u'summary',
  'These files are experimental, short lived, and the format will change.'),
 (u'product_version', '0.0.0'),
 (u'geospatial_lat_max', array([-31.])),
 (u'source',
  'This data is a reprojection and retile of Landsat surface reflectance scene data available from /g/data/rs0/scenes/'),
 (u'title',
  'Experimental Data files From the Australian Geoscience Data Cube - DO NOT USE'),
 (u'geospatial_lon_resolution', '0.00025'),
 (u'geospatial_lon_max', array([ 135.])),
 (u'Conventions', 'CF-1.6'),
 (u'history', 'NetCDF-CF file created 20151104.'),
 (u'geospatial_lat_min', array([-32.])),
 (u'geospatial_lat_resolution', '0.00025'),
 (u'date_created', '2015-11-04T15:45:28.399499'),
 (u'license', 'Creative Commons Attribution 4.0 International CC BY 4.0')]
>>> nc_ul = (fid.attrs['geospatial_lon_min'][0], fid.attrs['geospatial_lat_max'][0])
>>> nc_ur = (fid.attrs['geospatial_lon_max'][0], fid.attrs['geospatial_lat_max'][0])
>>> nc_lr = (fid.attrs['geospatial_lon_max'][0], fid.attrs['geospatial_lat_min'][0])
>>> nc_ll = (fid.attrs['geospatial_lon_min'][0], fid.attrs['geospatial_lat_min'][0])
>>> nc_ul
(134.0, -31.0)
>>> nc_ur
(135.0, -31.0)
>>> nc_lr
(135.0, -32.0)
>>> nc_ll
(134.0, -32.0)
>>> fid['longitude'][1:11] - fid['longitude'][0:10]
array([ 0.00025,  0.00025,  0.00025,  0.00025,  0.00025,  0.00025,
        0.00025,  0.00025,  0.00025,  0.00025])
```

The only thing might be the co-ordinate arrays which still reference the UL corner of a pixel.

``` python
>>> fid['longitude'][0:10]
array([ 134.     ,  134.00025,  134.0005 ,  134.00075,  134.001  ,
        134.00125,  134.0015 ,  134.00175,  134.002  ,  134.00225])
```

But CF is a little vague as to what it should be, so they'll be fine.  If they need to be changed it's a quick fix to add a 0.5 to the pixel indices during calculation.

They do confirm with an alternate method of calculation.

``` python
>>> gb = GriddedGeoBox(shape=(4000, 4000), origin=nc_ul)
>>> import numpy
>>> x, y = (numpy.arange(4000), numpy.arange(4000)) * gb.affine
>>> x
array([ 134.     ,  134.00025,  134.0005 , ...,  134.99925,  134.9995 ,
        134.99975])
>>> x - fid['longitude']
array([ 0.,  0.,  0., ...,  0.,  0.,  0.])
>>> (x - fid['longitude']).sum()
0.0
```

I'll close the issue.
